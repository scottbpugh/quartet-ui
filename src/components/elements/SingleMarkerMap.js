// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import React, {Component} from "react";
import "openlayers/dist/ol.css";

var ol = require("openlayers/dist/ol.js");

export class SingleMarkerMap extends Component {
  componentDidMount() {
    let delay = 0;
    if (this.props.delay) {
      delay = Number(this.props.delay);
    }
    window.setTimeout(() => {
      this.setUpMap();
      this.setMarkers(this.props.markerLocation);
      this._map.setTarget(this.props.targetId);
    }, delay);
  }
  componentWillUnmount() {
    this._map = null;
  }
  componentWillReceiveProps(nextProps) {
    if (JSON.stringify(this.props) !== JSON.stringify(nextProps)) {
      this.setMarkers(nextProps.markerLocation);
    }
  }
  setMarkers(markerLocation) {
    //create the style
    var iconStyle = new ol.style.Style({
      image: new ol.style.Icon(
        /** @type {olx.style.IconOptions} */ ({
          anchor: [0.5, 1],
          anchorOrigin: "bottom-right",
          scale: 0.4,
          anchorXUnits: "fraction",
          anchorYUnits: "pixels",
          opacity: 0.9,
          src: "https://unpkg.com/leaflet@1.3.1/dist/images/marker-icon-2x.png"
        })
      )
    });

    let f = new ol.Feature({
      geometry: new ol.geom.Point(
        ol.proj.fromLonLat(this.props.markerLocation)
      ),
      name: "Location" + this.props.targetId,
      id: "location" + this.props.targetId
    });

    let vectorSource = new ol.source.Vector({features: [f]});
    if (this._vectorLayer) {
      //this._map.removeLayer(this._vectorLayer);
    }

    let vectorLayer = new ol.layer.Vector({
      source: vectorSource,
      style: iconStyle
    });
    this._vectorLayer = vectorLayer;
    this._map.addLayer(this._vectorLayer);
    this._map.getView().fit(vectorSource.getExtent(), false);
  }
  setUpMap() {
    this._map = new ol.Map({
      layers: [new ol.layer.Tile({source: new ol.source.OSM()})],
      view: new ol.View({
        center: ol.proj.fromLonLat(this.props.markerLocation),
        zoom: 1,
        maxZoom: 12,
        minZoom: 1
      })
    });
  }
  render() {
    return (
      <div
        id={this.props.targetId}
        style={
          this.props.size ? this.props.size : {height: "150px", width: "200px"}
        }
      />
    );
  }
}
