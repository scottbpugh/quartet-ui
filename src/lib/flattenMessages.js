// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

export function flattenMessages(nestedMessages, prefix = "") {
  return Object.keys(nestedMessages).reduce((messages, key) => {
    const value = nestedMessages[key];
    const prefixedKey = prefix ? `${prefix}.${key}` : key;

    if (typeof value === "string") {
      messages[prefixedKey] = value;
    } else {
      Object.assign(messages, flattenMessages(value, prefixedKey));
    }

    return messages;
  }, {});
}

window.qu4rtet.exports("lib/flattenMessages", this);
