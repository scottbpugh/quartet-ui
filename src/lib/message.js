// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import {Position, Toaster, Intent} from "@blueprintjs/core";
import {pluginRegistry} from "plugins/pluginRegistration";
import stringHash from "string-hash";

const MINIMUM_EXPIRE = 7000;

const msgToaster = Toaster.create({
  className: "my-toaster",
  position: Position.BOTTOM_RIGHT
});

const getIntent = type => {
  switch (type) {
    case "success":
      return Intent.SUCCESS;
    case "danger":
    case "error":
      return Intent.DANGER;
    case "warning":
      return Intent.WARNING;
    default:
      return Intent.PRIMARY;
  }
};

const _msgHashMap = {};

const showMessageNoRepeat = (toastMessage, expires_in = MINIMUM_EXPIRE) => {
  let msgHash = stringHash(JSON.stringify(toastMessage));
  let timestamp = new Date().getTime();
  if (msgHash in _msgHashMap && _msgHashMap[msgHash].expires > timestamp) {
    // don't show msg and update the expiration to prevent the same message
    // over and over.
    _msgHashMap[msgHash].expires = timestamp + expires_in;
    return;
  }
  _msgHashMap[msgHash] = {expires: timestamp + expires_in};
  msgToaster.show(toastMessage);
};

export const showMessage = msg => {
  try {
    // for errors etc out of a component context.
    const intl = pluginRegistry.getIntl();
    if (msg.id) {
      showMessageNoRepeat(
        {
          message: intl.formatMessage({id: msg.id}, msg.values),
          intent: getIntent(msg.type)
        },
        msg.expires_in || MINIMUM_EXPIRE
      );
    } else if (msg.msg) {
      //let formatMsg = typeof msg.msg === "object" ? JSON.stringify(msg.msg) : msg.msg;
      showMessageNoRepeat(
        {message: msg.msg, intent: getIntent(msg.type)},
        msg.expires_in || MINIMUM_EXPIRE
      );
    }
  } catch (e) {
    console.log(
      "an error occurred while attempting to display an error toast.",
      e
    );
  }
};

window.qu4rtet.exports("lib/message", this);
