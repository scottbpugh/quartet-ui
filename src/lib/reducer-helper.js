// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// A helper function meant to make life a little easier in reducers.

/**
 * Unknown - Description
 *
 * @param {object} [state={}]       The reducer state
 * @param {type}   serverID         The serverID
 * @param {object} [mergeObject={}] The object to merge into the state for this server.
 *
 * @return {type} Description
 */
export const setServerState = (state = {}, serverID, mergeObject = {}) => {
  if (!serverID) {
    throw new Error(
      "A server ID must be defined to set the server state in reducer."
    );
  }
  const copy = {...state}; // use a copy of state for manipulations.
  if (!copy.servers) {
    // if servers are not created yet, create an object.
    copy.servers = {};
  }
  if (!copy.servers[serverID]) {
    // if there is no object for this server, create it.
    copy.servers[serverID] = {};
  }
  return {
    ...copy,
    servers: {
      ...copy.servers,
      [serverID]: {
        ...copy.servers[serverID],
        ...mergeObject
      }
    }
  };
};
window.qu4rtet.exports("lib/reducer-helper", this);
