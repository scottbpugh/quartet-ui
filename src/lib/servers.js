// Copyright (c) 2018 Serial Lab
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import Swagger from "swagger-client";
import {pluginRegistry} from "plugins/pluginRegistration";
import {store} from "store";
import actions from "actions/serversettings";

/**
 * Server - Holds data about a server and its settings, API client, ...
 */
export class Server {
  constructor(serverSettings) {
    /*
    Following object should be passed.
    {
      serverSettingName,
      hostname,
      port,
      ssl,
      serverID,
      path,
      username,
      password
    }
    */
    this.setServerData(serverSettings);
    // make saved server object available to core and plugins.
    pluginRegistry.registerServer(this);
  }

  setServerData = serverSettings => {
    /*
    Following object should be passed.
    {
      serverSettingName,
      hostname,
      port,
      ssl,
      serverID,
      path,
      username,
      password
    */
    // allow for partial updates.
    for (let key in serverSettings) {
      if (serverSettings[key] !== "" || serverSettings[key] !== undefined) {
        this[key] = serverSettings[key];
      }
    }
    this.protocol = this.ssl === true ? "https" : "http";
    this.port = !this.port ? 80 : this.port;
    this.path = !this.path && !serverSettings.path ? "" : this.path;
    this.url = this.getServerURL();
    this._client = null;
    this.manifest = null;
    this.appList = [];
  };
  getFormStructure = () => {
    return Server.getFormStructure(this.toJSON());
  };
  static getFormStructure = (initialValues = {}) => [
    {
      name: "serverSettingName",
      defaultValue: initialValues.serverSettingName || null,
      description: {
        type: "text",
        required: true,
        read_only: false,
        label: "Server Setting Name",
        help_text:
          "The label that will be used for this server connection setting."
      }
    },
    {
      name: "serverID",
      defaultValue: initialValues.serverID || null,
      description: {
        type: "hidden",
        required: false,
        read_only: false,
        label: "Server ID",
        help_text: "Hidden Server ID"
      }
    },
    {
      name: "hostname",
      defaultValue: initialValues.hostname || null,
      description: {
        type: "text",
        required: true,
        read_only: false,
        label: "Server Hostname",
        help_text:
          "A hostname or IP address, example localhost, serial-box.com, or 192.168.5.10."
      }
    },
    {
      name: "port",
      defaultValue: initialValues.port || null,
      description: {
        type: "number",
        required: true,
        read_only: false,
        label: "Port Number",
        min_value: 1,
        max_value: 65535,
        help_text: "A port to connect to. Example, 80, 8080, 443, ..."
      }
    },
    {
      name: "path",
      defaultValue: initialValues.path || "",
      description: {
        type: "text",
        required: false,
        read_only: false,
        label: "Root Path",
        helperText: "A path to interact with API (Optional), example /api"
      }
    },
    {
      name: "ssl",
      defaultValue: initialValues.ssl || false,
      description: {
        type: "boolean",
        required: false,
        read_only: false,
        label: "SSL/TLS",
        helperText: "SSL/TLS encryption"
      }
    },
    {
      name: "username",
      defaultValue: initialValues.username || null,
      description: {
        type: "text",
        required: false,
        read_only: false,
        label: "Username",
        help_text: "Basic Auth Username"
      }
    },
    {
      name: "password",
      defaultValue: initialValues.password || null,
      description: {
        type: "password",
        required: false,
        read_only: false,
        label: "Password",
        help_text: "Basic Auth Password"
      }
    }
  ];
  getArrayFields = () => {
    return [
      {name: "protocol", value: this.protocol, editable: true},
      {name: "port", value: this.port, editable: true},
      {name: "path", value: this.path, editable: true},
      {name: "ssl", value: this.ssl, editable: true},
      {name: "url", value: this.url, editable: false},
      {name: "hostname", value: this.hostname, editable: true},
      {
        name: "serverSettingsName",
        value: this.serverSettingName,
        editable: true
      },
      {name: "username", value: this.username},
      {name: "serverID", value: this.serverID}
    ];
  };

  toJSON() {
    return {
      serverID: this.serverID,
      protocol: this.protocol,
      port: this.port,
      path: this.path,
      ssl: this.ssl,
      hostname: this.hostname,
      serverSettingName: this.serverSettingName,
      url: this.url,
      appList: this.appList,
      username: this.username,
      password: this.password
    };
  }

  getServerURL = () => {
    return `${this.protocol}://${this.hostname}:${this.port}/${this.path}`;
  };

  parseSchema = () => {
    return new Promise((resolve, reject) => {
      Swagger(this.url + "schema/")
        .then(client => {
          // swagger-js client is available.
          // client.spec / client.originalSpec / client.errors
          resolve(client);
        })
        .catch(error => {
          reject(error);
        });
    });
  };

  listApps = () => {
    if (this.appList.length === 0) {
      this.getClient()
        .then(client => {
          this.appList = Object.keys(client.apis);
          // let redux know we got our data
          store.dispatch({
            type: actions.appsListUpdated,
            payload: this.toJSON()
          });
        })
        .catch(error => {
          // handle legacy.
          this.appList = ["serialbox"];
          // let redux know we got our data
          store.dispatch({
            type: actions.appsListUpdated,
            payload: this.toJSON()
          });
        });
    }
  };

  getClient = () => {
    return new Promise((resolve, reject) => {
      if (this._client) {
        resolve(this._client);
      } else {
        // we don't have a client yet.
        // Fetch it.
        this.parseSchema()
          .then(client => {
            this._client = client;
            resolve(this._client);
          })
          .catch(error => {
            reject(error);
          });
      }
    });
  };

  getManifest = () => {
    new Promise((resolve, reject) => {
      if (!this.manifest) {
        this.getClient().then(client => {
          client.apis.manifest.quartet_manifest_list().then(result => {
            this.manifest = result.body;
            resolve(result.body);
          });
        });
      } else {
        resolve(this.manifest);
      }
    });
  };
}
