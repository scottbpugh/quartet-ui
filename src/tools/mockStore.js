// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import React, {Component} from "react";

import {IntlProvider} from "react-intl";
import {addLocaleData} from "react-intl";
import en from "react-intl/locale-data/en";
import fr from "react-intl/locale-data/fr";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import {pluginRegistry} from "plugins/pluginRegistration";
import MockInitialState from "./mock-initial-state";
import {Provider} from "react-redux";
import {MemoryRouter as Router} from "react-router-dom";
import {injectIntl} from "react-intl";

addLocaleData([...en, ...fr]);
let defaultLocale = "en-US";
const middlewares = [thunk];

// Fake local storage.
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value;
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export const localStorage = new LocalStorageMock();
if (window && !window.localStorage) {
  // add fake local storage for tests.
  window.localStorage = localStorage;
}

if (window && !window.require) {
  // add fake require for ipcRenderer.
  window.require = module => {
    if (module === "electron") {
      return {
        ipcRenderer: {
          on: () => {},
          send: () => {}
        },
        remote: {
          app: {
            getVersion: () => {
              return "test-version";
            }
          }
        }
      };
    }
  };
}
/*
      remote: {
        app: {
          getVersion: function() {
            return "test-version";
          }
        }
*/
export const initialState = MockInitialState;

// Used by certain tests.
export const updateRegistryIntl = (locale, messages) => {
  const {intl} = new IntlProvider({
    locale: locale,
    messages: messages,
    defaultLocale: "en-US"
  }).getChildContext();
  pluginRegistry.registerIntl(intl);
};

export const mockStore = configureStore(middlewares);

// used to get a ref of the intl into our tests.
class _Wrapper extends Component {
  componentDidMount() {
    pluginRegistry.registerIntl(this.props.intl);
  }
  render() {
    return <div>{this.props.children}</div>;
  }
}
const Wrapper = injectIntl(_Wrapper);

// The TestWrapper to use for your tests.
export const TestWrapper = ({locale, messages, store, children}) => {
  if (!store) {
    // it uses the default "real app" store if no store is specified.
    store = require("store").store;
  }
  if (locale === undefined) {
    locale = defaultLocale;
  }
  if (messages === undefined) {
    messages = store.getState().intl.messages;
  }
  return (
    <IntlProvider locale={locale} messages={messages}>
      <Wrapper>
        <Provider store={store}>
          <Router>{children}</Router>
        </Provider>
      </Wrapper>
    </IntlProvider>
  );
};
