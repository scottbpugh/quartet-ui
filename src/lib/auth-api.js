// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {showMessage} from "lib/message";
import {getSyncValidators} from "components/elements/forms";
import {pluginRegistry} from "plugins/pluginRegistration";

/**
 * prepHeaders - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers.
 */
export const prepHeaders = (server, method = "GET") => {
  let headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  return {
    method: method,
    headers: headers,
    mode: "cors"
  };
};

/**
 * prepHeaders - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers and basic auth.
 */
export const prepHeadersAuth = async (server, method = "GET") => {
  let headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  headers.append(
    "Authorization",
    await pluginRegistry.getServer(server.serverID).getAuthorization()
  );
  return {
    method: method,
    headers: headers,
    mode: "cors"
  };
};

/**
 * getFormInfo - Description
 *
 * @param {A server object} server       Description
 * @param {Path to endpoint} path         Description
 * @param {Callback to set the state with formStructure} createForm   Description
 * @param {type} processField Description
 *
 */
export const getFormInfo = async (server, path, createForm, processField) => {
  return fetch(`${server.url}${path}`, await prepHeadersAuth(server, "OPTIONS"))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      // parse the values and filter to the one that are not readonly.
      let postFields = data.actions.POST;
      let formStructure = Object.keys(postFields)
        .map(field => {
          if (postFields[field].read_only === false) {
            return {name: field, description: postFields[field]};
          } else {
            return null;
          }
        })
        .filter(fieldObj => {
          if (processField) {
            return processField(fieldObj);
          }
          if (fieldObj) {
            // create sync validation arrays.
            fieldObj.validate = getSyncValidators(fieldObj);
            return true;
          }
          return false;
        });
      createForm(formStructure);
    })
    .catch(error => {
      showMessage({
        type: "danger",
        id: "app.servers.errorFormFetch",
        values: {error: error, serverName: server.serverSettingName}
      });
      throw error;
    });
};
