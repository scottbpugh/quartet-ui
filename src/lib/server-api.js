// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import {showMessage} from "lib/message";
import {getSyncValidators} from "components/elements/forms";
import {pluginRegistry} from "plugins/pluginRegistration";

/**
 * prepHeadersAuth - Prepares the headers to be sent.
 *
 * @param {object} server A server setting object
 *
 * @return {object} A request init object with headers and basic auth.
 */
export const prepHeadersAuth = async (server, method = "GET") => {
  const headers = new Headers();
  headers.append("Accept", "application/json");
  headers.append("Content-Type", "application/json");
  headers.append(
    "Authorization",
    await pluginRegistry.getServer(server.serverID).getAuthorization()
  );
  return {
    method,
    headers,
    mode: "cors"
  };
};

/**
 * getFormInfo - Returns an object with a form structure to be used by PageForm/redux form
 *
 * @param {A server object} server       Description
 * @param {Path to endpoint} path         Description
 * @param {Callback to set the state with formStructure} createForm   Description
 * @param {type} processField Description
 *
 */
export const getFormInfo = async (server, path, createForm, processField) => {
  return fetch(`${server.url}${path}`, await prepHeadersAuth(server, "OPTIONS"))
    .then(resp => {
      return resp.json();
    })
    .then(data => {
      // parse the values and filter to the one that are not readonly.
      try {
        const postFields = data.actions.POST;
        const formStructure = Object.keys(postFields)
          .map(field => {
            if (
              field === "id" ||
              (postFields[field].type === "field" &&
                postFields[field].read_only === true)
            ) {
              return null;
            }
            return {name: field, description: postFields[field]};
          })
          .filter(fieldObj => {
            if (processField) {
              return processField(fieldObj);
            }
            if (fieldObj) {
              // create sync validation arrays.
              fieldObj.validate = getSyncValidators(fieldObj);
              return true;
            }
            return false;
          });
        createForm(formStructure);
      } catch (e) {
        if (e.message.startsWith("Cannot read property 'POST'")) {
          pluginRegistry.getHistory().push("/access-denied");
        }
      }
    })
    .catch(error => {
      debugger;
      console.log("About to redirect to access denied");
      //pluginRegistry.getHistory().push("/access-denied");
      showMessage({
        type: "danger",
        id: "app.servers.errorFormFetch",
        values: {error, serverName: server.serverSettingName}
      });
      throw error;
    });
};

/**
 * fetchObject - fetches a single object from any Swagger API endpoint.
 *
 * @param {Server}   serverInstance  A Server instance
 * @param {string} [operationId=]  The Swagger operation id
 * @param {object} [parameters={}] Parameter passed with the call.
 *
 * @return {object} An object from the API call.
 */
export const fetchObject = async (
  serverInstance,
  operationId = "",
  parameters = {}
) => {
  const client = await serverInstance.getClient();
  try {
    const response = await client.execute({
      operationId,
      parameters,
      securities: {
        authorized: client.securities,
        specSecurity: [client.spec.securityDefinitions]
      }
    });
    if (response.ok) {
      return response.body;
    }
  } catch (e) {
    showMessage({
      type: "error",
      id: "plugins.numberRange.errorVanilla",
      values: {error: e}
    });
    throw e;
  }
};

/**
 * fetchPageList - Fetches a single list page.
 *
 * @param {type} serverInstance A Server instance
 * @param {type} operationId    The Swagger
 * @param {type} parameters     Parameters passed to the call (page...)
 *
 * @return {object} A response body
 */
export const fetchPageList = async (
  serverInstance,
  operationId,
  parameters
) => {
  const client = await serverInstance.getClient();
  try {
    const response = await client.execute({
      operationId,
      parameters,
      securities: {
        authorized: client.securities,
        specSecurity: [client.spec.securityDefinitions]
      }
    });
    if (response.ok) {
      return response.body;
    }
    throw new Error(response);
  } catch (e) {
    showMessage({
      type: "error",
      id: "plugins.numberRange.errorVanilla",
      values: {error: e}
    });
    throw e;
  }
};

/**
 * fetchListAll - Loops through all pages of a list and concats the result.
 *
 * @param {Server}   serverInstance  A Server Instance
 * @param {string} [operationId=]  Swagger operation id
 * @param {object} [parameters={}] Parameters passed to the call.
 * @param {array}  [results=[]]    The results, all objects for this API list.
 *
 * @return {type} Description
 */
export const fetchListAll = async (
  serverInstance,
  operationId = "",
  parameters = {},
  results = []
) => {
  return new Promise((resolve, reject) => {
    serverInstance
      .getClient()
      .then(client => {
        client
          .execute({
            operationId,
            parameters,
            securities: {
              authorized: client.securities,
              specSecurity: [client.spec.securityDefinitions]
            }
          })
          .then(response => {
            if (response.ok) {
              if (Array.isArray(response.body)) {
                results = results.concat(response.body);
                resolve(results);
              } else if (response.body.results) {
                results = results.concat(response.body.results);
                // pagination in effect. Assuming page number navigation.
                if (response.body.next) {
                  const url = new URL(response.body.next);
                  const page = new URLSearchParams(url.search).get("page");
                  const subParameters = {...parameters, page};
                  fetchListAll(operationId, subParameters, results)
                    .then(resolve)
                    .catch(reject);
                } else {
                  resolve(results);
                }
              }
            } else {
              reject(response);
            }
          })
          .catch(e => {
            reject(e);
          });
      })
      .catch(e => {
        reject(e);
      });
  });
};

/**
 * deleteObject - Deletes an object.
 *
 * @param {string} [operationId=]  The operationId to execute.
 * @param {object} [parameters={}] Object data
 *
 * @return {object} response.
 */
export const deleteObject = async (
  serverInstance,
  operationId = "",
  parameters = {}
) => {
  const client = await serverInstance.getClient();
  const response = await client.execute({
    operationId,
    parameters,
    securities: {
      authorized: client.securities,
      specSecurity: [client.spec.securityDefinitions]
    }
  });
  if (response.ok) {
    return response;
  }
};

window.qu4rtet.exports("lib/server-api", this);
