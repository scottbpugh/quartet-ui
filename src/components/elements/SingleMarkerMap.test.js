// Copyright (c) 2018 SerialLab Corp.
//
// GNU GENERAL PUBLIC LICENSE
//    Version 3, 29 June 2007
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
import React from "react";
import "tools/mockStore"; // mock ipcRenderer, localStorage, ...
import renderer from "react-test-renderer";
import {TestWrapper} from "tools/mockStore";
import Enzyme, {mount, shallow} from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import sinon from "sinon";
import toJson from "enzyme-to-json";
import * as serverAPI from "lib/server-api";
import {SingleMarkerMap} from "./SingleMarkerMap";

it("renders correctly", () => {
  const markerMap = renderer
    .create(
      <TestWrapper>
        <SingleMarkerMap
          size={{width: "260px"}}
          targetId={"map"}
          delay={0}
          markerLocation={[0.0, 10.0]}
        />
      </TestWrapper>
    )
    .toJSON();
  expect(markerMap).toMatchSnapshot();
});
