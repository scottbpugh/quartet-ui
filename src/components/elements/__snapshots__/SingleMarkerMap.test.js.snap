// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    id="map"
    style={
      Object {
        "width": "260px",
      }
    }
  />
</div>
`;
